utils.go

```diff
@@ -1,27 +1,122 @@
 package ipfs
 
 import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	ipfs "github.com/ipfs/go-ipfs-api"
+	cmds "github.com/ipfs/go-ipfs-cmds"
+
+	"github.com/beyondstorage/go-endpoint"
+	"github.com/beyondstorage/go-storage/v4/services"
 	"github.com/beyondstorage/go-storage/v4/types"
 )
 
 // Storage is the example client.
 type Storage struct {
+	ipfs *ipfs.Shell
+
 	defaultPairs DefaultStoragePairs
 	features     StorageFeatures
 
+	workDir string
+
 	types.UnimplementedStorager
+	types.UnimplementedCopier
+	types.UnimplementedMover
 }
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf("Storager IPFS {WorkDir: %s}", s.workDir)
 }
 
 // NewStorager will create Storager only.
 func NewStorager(pairs ...types.Pair) (types.Storager, error) {
-	panic("implement me")
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return nil, err
+	}
+
+	st := &Storage{
+		workDir: "/",
+	}
+	if opt.HasWorkDir {
+		st.workDir = opt.WorkDir
+	}
+
+	ep, err := endpoint.Parse(opt.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	e := "http://localhost:5001"
```

Comment:
Since `endpoint` is required, it seems not proper to set a default value here? If `ep` is invalid, will it be better that we report an error to the user instead of choosing a default value?

Reply:
Regarding the errors of global pairs, are there uniform definitions, or do I just use a custom error?

Reply:
https://github.com/beyondstorage/go-service-s3/blob/master/utils.go#L113 
Here `PairUnsupportedError` is used. What's your idea? And ping @JinnyYi to confirm.

Reply:
I agree with you.